src/fairseq2/nn/transformer/encoder.py

```diff
@@ -215,4 +245,24 @@ def extra_repr(self) -> str:
 
             s = f"{s}, self_attn_mask_factory={self_attn_mask_factory}"
 
+        if self.layer_drop_p > 0.0:
+            s = f"{s}, layer_drop_p={self.layer_drop_p:G}"
+
         return f"{s}, norm_order={self.norm_order}"
+
+
+# mypy: disable-error-code="no-any-return,override"
+
+
+def _record_drop_for_backward(x: Tensor, dropped_output: Tensor) -> Tensor:
+    return _RecordDropForBackwardFunction.apply(x, dropped_output)
+
+
+class _RecordDropForBackwardFunction(Function):
+    @staticmethod
+    def forward(ctx: Any, x: Tensor, dropped_output: Tensor) -> Tensor:
+        return x
+
+    @staticmethod
+    def backward(ctx: Any, grad_output: Tensor) -> Tuple[Tensor, Tensor]:
+        return grad_output, torch.zeros_like(grad_output)
```

Comment:
Smartly done! The gradient with respect to x is going to be just the same as the gradient of the output, since this is just the identity function when you drop layers.

Reply:
Thanks!

Reply:
Also, what is the advantage of using this method over PyTorch hooks?